instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/servlet/AppServerBridge.java

```diff
@@ -39,7 +40,15 @@ public static Context init(Context ctx) {
    * @return new context with AppServerBridge attached.
    */
   public static Context init(Context ctx, boolean shouldRecordException) {
-    return ctx.with(AppServerBridge.CONTEXT_KEY, new AppServerBridge(shouldRecordException));
+    Context context =
+        ctx.with(AppServerBridge.CONTEXT_KEY, new AppServerBridge(shouldRecordException));
+    // Add context for storing servlet context path. Servlet context path is updated from servlet
+    // integration.
+    if (context.get(ServletContextPath.CONTEXT_KEY) == null) {
+      context = context.with(ServletContextPath.CONTEXT_KEY, new AtomicReference<>(null));
+    }
+
+    return context;
```

Comment:
instead of adding an AtomicReference so that we can set the `contextPath` into the Context without having to create a new scope, would it work to modify the Servlet instrumentation so that it creates a new Scope with the `contextPath` as needed in the case where normal Servlet span is suppressed due to upstream app server span?

Reply:
I could add it in a separate advice
```
  public static class ServletContextPathAdvice {

    @Advice.OnMethodEnter(suppress = Throwable.class)
    public static void onEnter(
        @Advice.Argument(0) ServletRequest request,
        @Advice.Local("otelScope") Scope scope) {

      if (!(request instanceof HttpServletRequest)) {
        return;
      }

      int callDepth = CallDepthThreadLocalMap.incrementCallDepth(ServletContextPath.class);
      if (callDepth > 0) {
        return;
      }

      HttpServletRequest httpServletRequest = (HttpServletRequest) request;

      String contextPath = httpServletRequest.getContextPath();
      if (contextPath != null && !contextPath.isEmpty() && !contextPath.equals("/")) {
        Context context = Java8BytecodeBridge.currentContext().with(ServletContextPath.CONTEXT_KEY, contextPath);
        scope = context.makeCurrent();
      }
    }

    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
    public static void stop(
        @Advice.Local("otelScope") Scope scope) {
      if (scope != null) {
        CallDepthThreadLocalMap.reset(ServletContextPath.class);
        scope.close();
      }
    }
  }
```
and apply it to same methods as Servlet3Advice/Servlet2Advice. But now the question is where should I place this advice class. Easiest would be to copy paste it to both servlet2 & servlet3 but I doubt that you'll accept that. If I add it to servlet-common with a separate instrumentation module then tests that have `testInstrumentation project(":instrumentation:servlet:servlet-3.0:javaagent")` wouldn't use it because servlet-3.0 doesn't depend on servlet-common. I could either make servlet-3.0 depend on servlet-common or add servlet-common to tests that use `ServletContextPath.prepend` but that would create another issue. Dispatcher and response instrumentation in servlet-common create extra spans that would need handling in tests. Advice appreciated.

Reply:
I personally think that both `servlet-2.2` and `servlet-3.0` may depend on `servlet-common`. And in any case we should use all 2 (or even all 3) of them in tests. To be sure that they interplay correctly.